ecs_deploy/ecs.py

```diff
@@ -473,6 +595,267 @@ def apply_container_secrets(self, container, new_secrets, exclusive=False):
+        system_controls = {}
+        for system_control in system_controls_list:
+            system_controls.setdefault(system_control[0], [])
+
+            mapping = {}
+            mapping["namespace"] = system_control[1]
+            mapping["value"] = system_control[2]
+            system_controls[system_control[0]].append(mapping)
+            
+        self.validate_container_options(**system_controls)
+        for container in self.containers:
+            if container[u'name'] in system_controls:
+                self.apply_container_system_controls(
+                    container=container,
+                    new_system_controls=system_controls[container[u'name']],
+                    exclusive=exclusive,
+                )
+            elif exclusive is True:
+                self.apply_container_system_controls(
+                    container=container,
+                    new_system_controls={},
+                    exclusive=exclusive,
+                )
+
+    def apply_container_system_controls(self, container, new_system_controls, exclusive=False):
+        system_controls = container.get('systemControls', [])
+        old_system_controls = [ {name: value for name, value in system_control.items()} for system_control in system_controls]
+
+        if exclusive is True:
+            merged = new_system_controls if new_system_controls else []
+        else:
+            merged = copy.deepcopy(old_system_controls)
+            old_system_control_namespaces = [system_control["namespace"] for system_control in merged]
+            new_system_control_namespaces = [system_control["namespace"] for system_control in new_system_controls]
+            new = set(new_system_control_namespaces) - set(old_system_control_namespaces)
+            new_merged = list()
+            # update
+            for system_control in merged:
+                for new_system_control in new_system_controls:
+                    if system_control["namespace"] == new_system_control["namespace"]:
+                        system_control.update(new_system_control)
+            # new
+            for name in new:
+                for new_system_control in new_system_controls:
+                    if name == new_system_control["namespace"]:
+                        new_merged.append(new_system_control)
+
+            merged.extend(new_merged)
+
+        if old_system_controls == merged:
+            return
+
+        diff = EcsTaskDefinitionDiff(
+            container=container[u'name'],
+            field=u'systemControls',
+            value=merged,
+            old_value=old_system_controls
+        )
+        self._diff.append(diff)
+
+        container[u'systemControls'] = [
+            {name: value for name,value in e.items()} for e in merged
+        ]
+
+    def set_ulimits(self, ulimits_list, exclusive=False):
+        ulimits = {}
+        for ulimit in ulimits_list:
+            ulimits.setdefault(ulimit[0], [])
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

You are using the `dict.setdefault()` method. We recommend using the `defaultdict` object to handle missing items in internal state because the default value is automatically generated when you access or modify a missing key. `defaultdict` can also be easier to use and understand.

[Learn more](https://docs.python.org/3/library/collections.html#collections.defaultdict)

Reply:
Should be fine now.